### PR TITLE
Merge validated-entry into master to add ValidatedEntry widget

### PR DESCRIPTION
**Pull request by Dogeek**
### PR Details:
- Widget name: ValidatedEntry
- Author: Dogeek

### Description
Includes several ttk::Entry widgets that will validate the user input against specific patterns.

Includes :
* Validator classes to validate against regexes, passwords, integers, floats, percentages, emails etc
* Associated entry widgets for those validators
* Extensible architecture, creating a new validator is easy, and all the Entry widgets can be superseeded by other classes to implement easy validation
* Full test suite.

### Checklist
- [x] Widget in a separate file in the appropriate folder
- [x] Widget functions properly on both Windows and Linux
- [x] Widget code includes docstrings with parameter descriptions
- [x] Included an example file in `/examples`
- [x] Widget is covered by unitttests in `/tests`
- [x] Widget includes required assets files
- [x] Reference to widget in `AUTHORS.md`
- [x] Entry in sphinx documentation

### CODE REVIEW

File: ttkwidgets/validated_entries/validators.py
Comment (Dogeek) [reply ttkwidgets/validated_entries/validators.py:22]: > 
> 
> This is a widget with a lot of potential I think. I have a few questions, but if I am going out of the intended scope of the widget, please reel me back in.
> 
>     * There is no way that I can see that the programmer can request something of an `is_valid` property or something. Particularly for patterns like e-mails, this seems useful to me, as some input need not be valid when it is entered. For example, entering an e-mail address allows entering any data I like, it seems, requiring the programmer to do another check if it is a form or something like that.

This is a good idea, it's already pretty straightforward to access though, through the `self.entry.VALIDATOR._validate(text)` method. An `is_valid` property is very easy to implement, and I will implement it tonight, after work

>     * Maybe a visual cue could be implemented to indicate the validity of the input? Doing so on the entry itself is complex in `ttk`, especially if custom themes are used, but maybe a checkmark or something similar?

This widget is intended to be as barebones as possible. It's a validating entry, nothing more. Visual cues can be added by subclassing the widgets, but there is so many options for visual cues that it doesn't make sense to implement it in the widget itself.

>     * Separating the `Validator` and the `ValidatorEntry` feels a bit Java-style, but I understand the reasoning. Perhaps add an option for simply passing a regex to a `ValidatorEntry` to allow quick creation if a `Validator` need only be used once?

It is a bit Java-style but the intention is to have a clear separation between the data and the view, Integrating validation on the widget itself imo would blur the lines and make this whole thing less modular

>     * From my basic testing, it seems that `Validators` may not be combined or removed. Is this correct?

I intend to make it possible to have multiple validators for one widget (hence this particular architecture) and I will work on that tonight as well.

**Comment by Dogeek** [comment]
> 
> 
> This is a widget with a lot of potential I think. I have a few questions, but if I am going out of the intended scope of the widget, please reel me back in.
> 
>     * There is no way that I can see that the programmer can request something of an `is_valid` property or something. Particularly for patterns like e-mails, this seems useful to me, as some input need not be valid when it is entered. For example, entering an e-mail address allows entering any data I like, it seems, requiring the programmer to do another check if it is a form or something like that.

This is a good idea, it's already pretty straightforward to access though, through the `self.entry.VALIDATOR._validate(text)` method. An `is_valid` property is very easy to implement, and I will implement it tonight, after work

>     * Maybe a visual cue could be implemented to indicate the validity of the input? Doing so on the entry itself is complex in `ttk`, especially if custom themes are used, but maybe a checkmark or something similar?

This widget is intended to be as barebones as possible. It's a validating entry, nothing more. Visual cues can be added by subclassing the widgets, but there is so many options for visual cues that it doesn't make sense to implement it in the widget itself.

>     * Separating the `Validator` and the `ValidatorEntry` feels a bit Java-style, but I understand the reasoning. Perhaps add an option for simply passing a regex to a `ValidatorEntry` to allow quick creation if a `Validator` need only be used once?

It is a bit Java-style but the intention is to have a clear separation between the data and the view, Integrating validation on the widget itself imo would blur the lines and make this whole thing less modular

>     * From my basic testing, it seems that `Validators` may not be combined or removed. Is this correct?

I intend to make it possible to have multiple validators for one widget (hence this particular architecture) and I will work on that tonight as well.

File: docs/source/authors.rst
Comment (Dogeek) [reply docs/source/authors.rst:40]: noted, will fix

File: docs/source/ttkwidgets/ttkwidgets.validated_entries/ttkwidgets.validated_entries.validators.rst
Comment (Dogeek) [reply docs/source/ttkwidgets/ttkwidgets.validated_entries/ttkwidgets.validated_entries.validators.rst:32]: No reason, I just don't work with sphinx very much so I'm bound to make a few mistakes

File: ttkwidgets/validated_entries/validated_entry.py
Comment (Dogeek) [reply ttkwidgets/validated_entries/validated_entry.py:1]: Does it show that I hacked it together ? ahah

File: docs/source/authors.rst
Comment (Dogeek) [reply docs/source/authors.rst:17]: I'm using VS Code, it does trim whitespace on save, might be that I also wrote the doc entries really quickly

File: ttkwidgets/validated_entries/validators.py
Comment (Dogeek) [reply ttkwidgets/validated_entries/validators.py:22]: It does belong to the validator though, because the validator is responsible for generating the appropriate dict for the entry, not to mention that this is already a kwarg when instanciating the entry. I can absolutely check that kwargs are not superseeded by the validator though.

Comment (Dogeek) [reply ttkwidgets/validated_entries/validators.py:30]: You can pass in the 'validate' kwarg to the entry to overwrite the validate_on attribute of the validator.

File: docs/source/documentation.rst
Comment (RedFantom) [inline docs/source/documentation.rst:2]: So, for some reason the documentation on the ReadTheDocs build is not working, but turns out that this is an issue for all branches except the latest master documentation. The documentation matches the required format and builds perfectly fine locally, so I expect it will also do so on the ReadTheDocs once merged.

File: tests/test_validatedentries.py
Comment (RedFantom) [inline tests/test_validatedentries.py:10]: I think the name of this test should be changed?

File: tests/test_validatedentries.py
Comment (RedFantom) [inline tests/test_validatedentries.py:143]: Please add a newline to the end of the file.

File: ttkwidgets/__init__.py
Comment (RedFantom) [inline ttkwidgets/__init__.py:14]: Could you change either this or the `ttkwidgets/validated_entries/__init__.py` so that only the classes defined in the `validators.py` module are imported?

File: tests/test_validatedentries.py
Comment (Dogeek) [reply tests/test_validatedentries.py:10]: True, I copy the existing tests to get the boilerplate out of the way, I guess I forgot to rename the class

File: ttkwidgets/__init__.py
Comment (Dogeek) [reply ttkwidgets/__init__.py:14]: Can I ask why. I'd argue that the numbers and strings modules are more important since they define actual ready to use entries. If a dev wants to make a custom validator, he can import it with `from ttkwidgets.validated_entries.validators import *` right ?

Comment (RedFantom) [reply ttkwidgets/__init__.py:14]: While that is true, it introduces name shadowing, possible import loops (very unlikely in this case) and messes with editors that optimize imports. It hides behaviour that the programmer should decide what to do with, making code harder to debug. Especially because it's standard library modules, I really think it's better not to use an `import *`.

**Review by RedFantom — APPROVED** [review]
I still really like this widget. I understand your reasoning for not adding a visual cue, and I think you're right. Thank you for making the changes I requested, I think it's good to go now! 👍🏻

**Pull request merged**
